docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go

```diff
@@ -0,0 +1,180 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// +kubebuilder:docs-gen:collapse=Apache License
+
+/*
+We'll start out with some imports.  You'll see below that we'll need a few more imports
+than those scaffolded for us.  We'll talk about each one when we use it.
+*/
+package v1
+
+import (
+	"github.com/robfig/cron"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apimachineryutilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+/*
+Next, we'll setup a logger for the webhooks.
+*/
+
+var log = logf.Log.WithName("cronjob-resource")
+
+/*
+Notice that we use kubebuilder markers to generate webhook manifests.
+The first and second markers are responsible for generating a mutating webhook manifest and a validating
+webhook manifest respectively.
+
+The meaning of each marker can be found [here](../TODO.md).
+*/
+
+// +kubebuilder:webhook:path=/mutate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=true,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=mcronjob.kb.io
+// +kubebuilder:webhook:path=/validate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=false,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=vcronjob.kb.io
+
+/*
+We use the `webhook.Defaulter` interface to set defaults to our CRD.
+A webhook will automatically be served that calls this defaulting.
+*/
+
+var _ webhook.Defaulter = &CronJob{}
+
+/*
+The Default method is expected to mutate the receiver, setting the defaults.
+*/
+func (c *CronJob) Default() {
+	log.Info("defaulting cronjob", "namespace", c.Namespace, "name", c.Name)
+
+	if c.Spec.ConcurrencyPolicy == "" {
+		c.Spec.ConcurrencyPolicy = AllowConcurrent
+	}
+	if c.Spec.Suspend == nil {
+		c.Spec.Suspend = new(bool)
+	}
+	if c.Spec.SuccessfulJobsHistoryLimit == nil {
+		c.Spec.SuccessfulJobsHistoryLimit = new(int32)
+		*c.Spec.SuccessfulJobsHistoryLimit = 3
+	}
+	if c.Spec.FailedJobsHistoryLimit == nil {
+		c.Spec.FailedJobsHistoryLimit = new(int32)
+		*c.Spec.FailedJobsHistoryLimit = 1
+	}
+}
+
+/*
+We use the `webhook.Validator` interface to validate our CRD.
```

Comment:
`to validate our CRD beyond what's possible with declarative validation.  Generally, declarative validation should be sufficient, but sometimes more advanced usecases call for complex validation` or something

Reply:
For instance, we'll see below that we use this to validate a well-formed cron schedule without making up a long regular expression.